server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
 import org.uiautomation.ios.wkrdp.message.WebkitApplication;
 import org.uiautomation.ios.wkrdp.message.WebkitPage;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.logging.Level;
-import java.util.logging.Logger;
-
 // TODO freynaud delete the interface ?
 public class WebKitNotificationListener implements MessageListener {
 
   private static final Logger log = Logger.getLogger(WebKitNotificationListener.class.getName());
+
   private final RemoteIOSWebDriver driver;
+
   private final ServerSideSession session;
+
   private final WebKitSynchronizer sync;
 
-  public WebKitNotificationListener(RemoteIOSWebDriver driver, WebKitSynchronizer syncronizer,
-                                    ServerSideSession session) {
+  private List<WebkitApplication> registeredApplications;
+
+  private Lock applicationRegistrationLock;
+
+  private Lock pagesProcessLock;
+
+  public WebKitNotificationListener(RemoteIOSWebDriver driver, WebKitSynchronizer syncronizer, ServerSideSession session) {
     this.driver = driver;
     this.session = session;
     this.sync = syncronizer;
+    this.registeredApplications = new ArrayList<>();
+    this.applicationRegistrationLock = new ReentrantLock();
+    this.pagesProcessLock = new ReentrantLock();
   }
 
   @Override
   public void onMessage(IOSMessage message) {
     if (message instanceof ReportSetupMessage) {
-      ReportSetupMessage m = (ReportSetupMessage) message;
-      driver.setDevice(m.getDevice());
-      sync.signalSimRegistered();
+      ReportSetupMessage reportSetupMessage = ReportSetupMessage.class.cast(message);
+      handle(reportSetupMessage);
+    } else if (message instanceof ReportConnectedApplicationsMessage) {
+      ReportConnectedApplicationsMessage reportConnectedApplicationsMessage = ReportConnectedApplicationsMessage.class
+          .cast(message);
+      handle(reportConnectedApplicationsMessage);
+    } else if (message instanceof ApplicationSentListingMessage) {
+      ApplicationSentListingMessage applicationSentListingMessage = ApplicationSentListingMessage.class.cast(message);
+      handle(applicationSentListingMessage);
+    } else if (message instanceof ApplicationDataMessage) {
+      // ApplicationDataMessage m = (ApplicationDataMessage)message;
```

Comment:
@krmahadevan 
Yes it did not have any implementation earlier. I did not know what to do with it. I can remove it if needed.